### PR TITLE
[deckhouse] update mpo to work with the new module system

### PR DESCRIPTION
**Pull request by ipaqsa**
## Description
It deprecates v1alpha1 for module pull override resource, and provides v1alpha2.

## Why do we need it, and what problem does it solve?
The current version of module pull override contains the source field, it is not met our new model.

Now, by MPO you can only select a tag of the module, but the source is set by the module config controller, so MPO requires a module to be enabled and to have an active source. 

Flow to create MPO:
1. Create a module source.
2. Enable a module(if conflict solve it) by the module config.
2.1. Select the source by the module config, if there are several sources.
3. Create a module pull override.

Flow to change the source for a module with MPO:
1. Change source in module config
2. Wait for resync

## What is the expected result?

Creating MPO for a non-existing module:

```
root@dev-master-0:~# kubectl get modulepulloverrides.deckhouse.io 
NAME    UPDATED   MESSAGE
echho   10s       The module not found
```

Creating MPO for a module that does have source:

```
root@dev-master-0:~# kubectl get modulepulloverrides.deckhouse.io 
NAME    UPDATED   MESSAGE
echo    12s       The module does not have an active source
```

Creating MPO for a module with non-existing source:

```
root@dev-master-0:~# kubectl get modulepulloverrides.deckhouse.io 
NAME    UPDATED   MESSAGE
echo    12s       The source not found
```

Creating MPO for a disabled module:

```
root@dev-master-0:~# kubectl get modulepulloverrides.deckhouse.io 
NAME   UPDATED   MESSAGE
echo    7s       The module disabled
```

Creating MPO with wrong tag:

```
root@dev-master-0:~# kubectl get mpo
NAME   UPDATED   MESSAGE
echo   4s        Download error: GET https://dev-registry.deckhouse.io/v2/deckhouse/losev/external-modules/echo/manifests/tmp-hook-error2: MANIFEST_UNKNOWN: manifest unknown; map[Tag:tmp-hook-error2]
```

Creating MPO for a normal module:

```
root@dev-master-0:~# kubectl get modulepulloverrides.deckhouse.io 
NAME    UPDATED   MESSAGE
echo    15s       Ready
```

Deleting a module config when MPO exists:

```
root@dev-master-0:~# kubectl delete mc echo 
Error from server: admission webhook "module-configs.deckhouse-webhook.deckhouse.io" denied the request: delete ModulePullOverride before deleting module config
```

A module with MPO has a new condition overridden:
```
apiVersion: deckhouse.io/v1alpha1
kind: Module
metadata:
  creationTimestamp: "2024-11-18T15:34:15Z"
  generation: 16
  labels:
    deckhouse.io/epoch: "1326105356"
  name: echo
  resourceVersion: "230347744"
  uid: 7111cee7-50cd-4ecf-ba20-d691b13b0f59
properties:
  availableSources:
  - losev-test
  releaseChannel: Stable
  requirements:
    deckhouse: '> v1.63.0'
    kubernets: '> v1.30.0'
  source: losev-test
  weight: 910
status:
  conditions:
  - lastProbeTime: "2024-12-03T15:57:20Z"
    lastTransitionTime: "2024-12-03T15:57:20Z"
    status: "True"
    type: EnabledByModuleConfig
  - lastProbeTime: "2024-12-03T15:59:58Z"
    lastTransitionTime: "2024-12-03T15:57:26Z"
    status: "True"
    type: EnabledByModuleManager
  - lastProbeTime: "2024-12-03T15:59:58Z"
    lastTransitionTime: "2024-12-03T15:56:23Z"
    message: 'module hook ''dev/hooks/hello-on-startup.py'' failed: exit status 1'
    reason: ModuleError
    status: "False"
    type: IsReady
  - lastProbeTime: "2024-12-03T15:59:48Z"
    lastTransitionTime: "2024-12-03T15:56:47Z"
    status: "True"
    type: IsOverridden
  phase: Error
```

Set version from MPO at restart:

Version from release:
```
root@dev-master-0:~# kubectl get module echo -owide
NAME   WEIGHT   RELEASE CHANNEL   SOURCE       VERSION   PHASE   ENABLED   DISABLED MESSAGE   READY
echo   925      Stable            losev-test   v0.7.26   Ready   True                         True
```

Version after MPO:
```
root@dev-master-0:~# kubectl get module echo -owide
NAME   WEIGHT   RELEASE CHANNEL   SOURCE       VERSION          PHASE           ENABLED   DISABLED MESSAGE   READY
echo   910      Stable            losev-test   tmp-hook-error   HooksDisabled   True                         False
```

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: feature 
summary: Update mpo to work with the new module system. Deprecate v1alpha1 for ModulePullOverride.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Taior. I've started the workflow for label 'e2e/run/aws'...

**Comment by Taior** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.29 12198250894 12198250894-1-con-1-29 /sys/deckhouse-oss/install:pr10903 user@34.159.185.85

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Taior. I've recognized your '/e2e/abort' command and started the workflow...

File: deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_module_config.go [hunk 1]
```diff
@@ -79,6 +80,14 @@ func moduleConfigValidationHandler(cli client.Client, moduleStorage moduleStorag
 					}
 				}
 
+				exist, err := utils.ModulePullOverrideExists(ctx, cli, cfg.Name)
+				if err != nil {
+					return nil, fmt.Errorf("get the '%s' module pull override: %v", cfg.Name, err)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_module_config.go:85]:
```suggestion
					return nil, fmt.Errorf("get the '%s' module pull override: %w", cfg.Name, err)
```

File: deckhouse-controller/pkg/apis/deckhouse.io/v1alpha2/module_pull_override.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2023 Flant JSC
```
Comment (miklezzzz) [inline deckhouse-controller/pkg/apis/deckhouse.io/v1alpha2/module_pull_override.go:2]:
```suggestion
Copyright 2024 Flant JSC
```

File: deckhouse-controller/pkg/controller/module-controllers/override/controller.go [hunk 3]
```diff
@@ -126,198 +129,295 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		controllerutil.RemoveFinalizer(mpo, v1alpha1.ModulePullOverrideFinalizer)
+		if err = r.client.Update(ctx, mpo); err != nil {
+			r.log.Errorf("failed to remove finalizer for the '%s' module pull override: %v", mpo.Name, err)
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+	return ctrl.Result{Requeue: true}, nil
```
Comment (miklezzzz) [inline deckhouse-controller/pkg/controller/module-controllers/override/controller.go:420]: is requeuing a deleted mpo required?

Comment (ipaqsa) [reply deckhouse-controller/pkg/controller/module-controllers/override/controller.go:420]: No

Comment (ipaqsa) [reply deckhouse-controller/pkg/controller/module-controllers/override/controller.go]: fixed

File: deckhouse-controller/pkg/controller/module-controllers/override/controller.go [hunk 4]
```diff
@@ -126,198 +129,295 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		if uerr := r.updateModulePullOverrideStatus(ctx, mpo); uerr != nil {
+			r.log.Errorf("failed to update the '%s' module pull override: %v", mpo.Name, uerr)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{RequeueAfter: defaultRequeueAfter}, nil
+	}
+
+	// source must be
+	if module.Properties.Source == "" {
+		r.log.Debugf("the '%s' module does have an active source, skip it", mpo.Name)
```
Comment (miklezzzz) [inline deckhouse-controller/pkg/controller/module-controllers/override/controller.go:195]:
```suggestion
		r.log.Debugf("the '%s' module does not have an active source, skip it", mpo.Name)
```

**Comment by ipaqsa** [comment]
@yalosev @ldmonster @miklezzzz when we should use error ?
```
r.log.Errorf("failed to update the '%s' module pull override: %v", mpo.Name, uerr)
```

or 
```
r.log.Warnf("failed to update the '%s' module pull override: %v", mpo.Name, uerr)
```

File: deckhouse-controller/pkg/controller/moduleloader/sync.go [hunk 5]
```diff
@@ -72,7 +109,7 @@ func (l *Loader) restoreAbsentModulesFromOverrides(ctx context.Context) error {
 		}
 
 		// if deployedOn annotation isn't set or its value doesn't equal to current node name - overwrite the module from the repository
-		if annotationNodeName, set := mpo.GetAnnotations()[v1alpha1.ModulePullOverrideAnnotationDeployedOn]; !set || annotationNodeName != currentNodeName {
+		if deployedOn, set := mpo.GetAnnotations()[v1alpha1.ModulePullOverrideAnnotationDeployedOn]; !set || deployedOn != currentNodeName {
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/moduleloader/sync.go:112]: Is v1alpha1.MPO expected here?

Comment (yalosev) [reply deckhouse-controller/pkg/controller/moduleloader/sync.go:112]: oh, i got it. It's just a const

**Comment by miklezzzz** [comment]
it seems almost all controllers' errors should be converted to warnings, as there isn't much logic behind them (it's mostly about some unmet conditions and kube-api being unavailable), am I wrong? If a module can't be installed/updated/deleted, I barely need to traverse the stacktrace as I can pinpoint the issue merely by the message.

**Review by ldmonster — APPROVED** [review]

File: deckhouse-controller/crds/module-pull-override.yaml [hunk 6]
```diff
@@ -71,3 +72,55 @@ spec:
...
+        openAPIV3Schema:
+          type: object
+          description: |
+            Defines the configuration.
+          required:
+            - spec
+          properties:
+            spec:
+#             TODO(ipaqsa): Remove it after delete v1alpha1
+              x-kubernetes-preserve-unknown-fields: true
```
Comment (yalosev) [inline deckhouse-controller/crds/module-pull-override.yaml:88]:
```suggestion
```

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/openstack'...

**Pull request merged**